packages/react-native-sdk/src/hooks/useCallControls.tsx

```diff
@@ -54,33 +57,78 @@ export const useCallControls = () => {
 
   const publishVideoStream = useCallback(async () => {
     try {
-      if (currentVideoDevice) {
+      if (currentVideoDevice && isOnlineRef.current) {
         const videoStream = await getVideoStream(currentVideoDevice.deviceId);
-        if (call) await call.publishVideoStream(videoStream);
+        if (call) {
+          await call.publishVideoStream(videoStream);
+        }
       }
     } catch (e) {
       console.log('Failed to publish video stream', e);
     }
   }, [call, currentVideoDevice]);
 
-  /* Attempt to republish video stream when app comes back to foreground */
-  useAppStateListener(publishVideoStream);
+  const isAudioPublished = localParticipant?.publishedTracks.includes(
+    SfuModels.TrackType.AUDIO,
+  );
+  const isVideoPublished = localParticipant?.publishedTracks.includes(
+    SfuModels.TrackType.VIDEO,
+  );
+  /** Refs to keep track of whether the user has published the track before going offline  */
+  const isAudioPublishedRef = useRef(false);
+  const isVideoPublishedRef = useRef(false);
+  isAudioPublishedRef.current = !isAudioPublished;
+  isVideoPublishedRef.current = !isVideoPublished;
+
+  /** Refs to be used for useEffect that does the rejoining flow when coming back offline */
+  const publishAudioStreamRef = useRef(publishAudioStream);
+  const publishVideoStreamRef = useRef(publishVideoStream);
+  publishAudioStreamRef.current = publishAudioStream;
+  publishVideoStreamRef.current = publishVideoStream;
+
+  /** Attempt to republish video stream when app comes back to foreground */
+  useAppStateListener(
+    isVideoPublishedRef.current ? publishVideoStream : undefined,
+  );
+
+  /**
+   * Effect to re-join to an existing call happens in case the user comes back online
+   */
+  useEffect(() => {
+    const unsubscribe = NetInfo.addEventListener(async (state) => {
+      const { isConnected, isInternetReachable } = state;
+      const isOnline = isConnected !== false && isInternetReachable !== false;
+      isOnlineRef.current = isOnline;
+      if (!callRef.current) return;
+      const callToJoin = callRef.current;
+      await rejoinCall(
+        callToJoin,
+        isOnline,
+        isAudioPublishedRef.current,
+        isVideoPublishedRef.current,
+        publishAudioStreamRef.current,
+        publishVideoStreamRef.current,
+      );
+    });
+
+    return unsubscribe;
+  }, []);
```

Comment:
This effect should live somewhere else, we will have integrators that won't use this hook (when building their own call controls etc, or for any other reason)

Reply:
As discussed, [I created a card to tackle this later](https://www.notion.so/stream-wiki/Refactor-publish-functions-mute-state-and-subsequently-call-rejoin-effect-c792cdcee7ad401885723146e9e1a93b?pvs=4)